### ISSUE TITLE
[react-redux] injecting props HOC and forced exact type of OwnProps

### ISSUE THREAD

**Issue by jacekkolasa** [flow-typed/flow-typed#3072]
Currently OwnProps have to be Exact Type. Is it possible to use current connect<....> typings to create injecting-props HOC, like `injectProp` example in HOC docs?

https://flow.org/en/docs/react/hoc/
```
function injectProp<Config: {}>(
  Component: React.AbstractComponent<Config>
): React.AbstractComponent<$Diff<Config, {foo: number | void}>> {
  return function WrapperComponent(
    props: $Diff<Config, {foo: number | void}>,
  ) {
    return <Component {...props} foo={42} />;
  };
}
```
this doesn't work and complains about `OwnProps is incompatible with exact type`
```
type Props = { ...OwnProps, foo: number };
type OwnProps = {};
const mapStateToProps = () => ({ foo: 5 });
export default connect<Props, OwnProps, _, _, _, _>(mapStateToProps);
```
(cc @peter-leonov)

**Comment by peter-leonov** [comment]
Quick guess:

```js
-type OwnProps = {};
+type OwnProps = {||};
```

**Comment by jacekkolasa** [comment]
@peter-leonov yes, exactly. I cannot do `type OwnProps = {||};` because then i cannot pass any other props into the decorated component, like:
`let usage = <DecoratedConnectedComponent propToBePassedThru="bar" />;` - here i want `ConnectedComponent` to use `propToBePassedThru` but `injectProp` should not know about it

**Comment by peter-leonov** [comment]
This little test works fine:

```js
// @flow
import * as React from "react";
import { connect } from "react-redux";

export let e = [];

function checkSimplePropertyInjection() {
  type OwnProps = {|
    foo: number,
    bar: string,
  |};
  type Props = { ...OwnProps, foo: number };
  const mapStateToProps = () => ({ foo: 5 });

  class Com extends React.Component<Props> {}

  const Connected = connect<Props, OwnProps, _, _, _, _>(mapStateToProps)(Com);

  <Connected foo={42} bar="str" />;
  //$ExpectError property `foo` is missing in props [1] but exists in `OwnProps`
  <Connected bar="str" />;
  e.push(Connected);

  function injectProp<Config: {}>(
    Component: React.AbstractComponent<Config>,
  ): React.AbstractComponent<$Diff<Config, { foo: number | void }>> {
    return function WrapperComponent(
      props: $Diff<Config, { foo: number | void }>,
    ) {
      return <Component {...props} foo={42} />;
    };
  }

  const Decorated = injectProp(Connected);
  // OK without `foo`
  <Decorated bar="str" />;
  // OK with a not needed `foo`
  <Decorated foo={42} bar="str" />;
  //$ExpectError property `bar` is missing in props [3] but exists in `Props` [4]
  <Decorated />;
  e.push(Decorated);
}
```

**Comment by peter-leonov** [comment]
> I cannot do `type OwnProps = {||};`

True. In the little test above I had to add the `Connected` props to `OwnProps` to pass them from outside:

```js
  type OwnProps = {|
    foo: number,
    bar: string,
  |};
```

**Comment by jacekkolasa** [comment]
Big thanks for your quick replies. So the idea is to have the `connect()` work in separate file, pluggable to different components. Sometimes it is used along with other hocs, like (using recompose)
```
export default compose(
  injectProp, // uses connect
  withOtherHOC,
)
```
- that's why i want to have the `connect` fn separated

**Comment by peter-leonov** [comment]
Aha, this is more interesting then. I've never tried to compose `compose()`. Can you please help with making a minimal representative code example which does not work as expected (or at all 😅)?

**Comment by jacekkolasa** [comment]
The most minimal example is the one i've included earlier:
```
type Props = { ...OwnProps, foo: number };
type OwnProps = {};
const mapStateToProps = () => ({ foo: 5 });
export default connect<Props, OwnProps, _, _, _, _>(mapStateToProps);
```
Unfortunately i have problems with recreating a scenario with `compose()` where the error is caused by react-redux typings... Because some `recompose` hocs are not working properly with React.AbstractComponent (like in https://flow.org/en/docs/react/hoc/) so there is more than source of issues here.

Anyway, i basically wanted to know if OwnProps really have to be `exact` and you answered that question, thanks! I don't know if issues i'm encountering are solvable now. Bearing in mind that part of community is moving to `hooks` and `flow` is still changing, i doubt that the typing patterns for HOCs will be unified (react-redux, recompose, internal react-flowlib). Solutions you've provided are great for connect() functions which are bound to individual components.

I'm encountering so many errors when trying to compose HOCs so i don't even know where to start :) And it's hard for me to provide for you an example which will have isolated issue with `react-redux` only. For example, i can provide typings to following block by multiple different ways, but each way have some problems:
```
import { connect } from 'react-redux';
import { compose, withProps } from 'recompose';
const composed = compose(
  withProps(() => ({
    foo: 5,
  })),
  connect(() => ({ bar: 5 }))
)
export default composed;
```

**Comment by peter-leonov** [comment]
Thank you for providing this very interesting case while I'm still on vacation 😎 As far as I understand, the aim is to make result of `connect()` call composable or, at least, usable indirectly. Here I can prove that current typings work in a case without exporting the result, so there is some hope 😅:

```js
  declare var compose: $Compose;

  const composedDecorators = compose(
    connect(mapStateToProps),
    a => a,
  );

const Decorated = composedDecorators(Com);
//$ExpectError property `bar` is missing
<Decorated />;
```

Will look into the more general case shortly! Let the Flow be with us 🤓

**Comment by peter-leonov** [comment]
@jacekkolasa here is a PR (#3076) with the test cases as I see them and — wait for it! — a working support for inexact `OwnProps` 🎉 Hope it will still work with the next Flow version 🤞

**Comment by peter-leonov** [comment]
> The example is unfortunately not working in my case

Maybe it's because of my bad copy-paste skill :) Please, try using the [code from the PR](https://github.com/flow-typed/flow-typed/pull/3076/files#diff-75f78449d157eb85dd31fe88453d5ad2R44) directly. It passes unit tests for types, even in Flow 0.90 😎

> I don't know if that is in the same scope, but i wanted to share one concern - I don't know how recompose typings (which are old) do that, but for some reasons i don't have to write annotations when using `withProps` recompose function. On the other hand, to use `withProps` i have to use `HOC<...>` type exported by `recompose`
> - it doesn't work with React.AbstractComponent :(

As far as I understood, the pre `React.AbstractComponent` type used everywhere `React.ComponentType` is now an alias `type ComponentType<T> = AbstractComponent<T, any>`. So, theoretically, it should be fine if other libraries internally used the proper types. But I'm not 100% sure.

> Having the ability to just write one type annotation for the whole `compose` function makes it easier to work with. Imagine having 10 HOCs inside `compose()` function and having to annotate all the in/out props for every HOC

This is 100% true. I guess, Flow should allow this because (as far as I understand) it just needs the very tip of the exported type to be declared. If you get any success making it work, please, share 🤓

**Comment by peter-leonov** [comment]
IMHO, the idea to specify all the type arguments for each and every composed decorator looks fine to me. Otherwise I'd need to calculate all those `$Diff<>` and `{| ...OwnProps, ...StateProps |}` tricky type inference in my head and type the final function manually and statically. Once I need to make any change to any link in the type chain I'd need to recalculate it again. And Flow is not the very best helper here with it's (still some times) cryptic error messages.

But what I've found is that in case of the `connect(...)` is not the first argument to `compose()` it does not require the type arguments:

```js
const composedDecorators = compose(
  injectProp,
  connect(mapStateToProps),
);
export const Decorated = composedDecorators(Com);
// an error did not get caught:
<Decorated />;
```

...but, it does not report type errors as it should with type arguments provided:

```js
const composedDecorators = compose(
  injectProp,
//^^^^^^^^^^ bla-bla-bla … property `own1` is missing in  props [1] but exists in  `OwnProps` [2]
  connect<Props, OwnProps, _, _, _, _>(mapStateToProps),
);
export const Decorated = composedDecorators(Com);
// it does not report the error here but it does it above
<Decorated />;
```

So, again imho, it is better to specify type arguments for HOC as this makes Flow more stable and not swallowing important type conflicts.

**Comment by jacekkolasa** [comment]
After your changes i can successfully write the annotations as i described in first post, awesome! :)

You're right that if `connect` is not the first arg it behaves differently - but in my case if i do `const injectProp = (x) => x` then i still have error about missing annotations in connect - which is completely ok for me.

I found another error - if i try to export the composed function then i have to annotate `compose` as well. I tried to do this by the book, using example from official flow-hoc docs, but it looks like it's incompatible:
```
type OwnProps = {};
type Props = { ...OwnProps, foo: number };
const mapStateToProps = () => ({ foo: 5 });
type Composed = <Config: {}>(Component: React.ComponentType<Config>) =>
  React.ComponentType<$Diff<Config, { foo: number | void }>>;
const composed: Composed = compose(
  connect<Props, OwnProps, _, _, _, _>(mapStateToProps)
);
export default composed;
```
`` `Props` [1] is incompatible with `Config` [2].``
where Config is `React$AbstractComponent` definition from flowlib/react.js


About annotating every decorator in `compose()` chain - if you ask me i just want it to work :) I would prefer Flow to detect internal typings but if that's hard, then it's ok to type manually individual decorators. For example - `withProps` decorator from `recompose` works ok without annotations but we have to admit it's much simpler HOC than `connect()`...

**Comment by peter-leonov** [comment]
First of all please try defining `Props` type like this:
```js
type Props = { ...$Exact<OwnProps>, foo: number };
```
this will fix a very weird bug with some `undefined` being not compatible with blah.

Second idea is that `type Composed` may be just defined incorrectly 😝 If we check the test from the PR, it uses `AbstractComponent` also with some `Config` and it kinda works magically. Hope this helps 😌

**Comment by jacekkolasa** [comment]
Thanks for the tip, unfortunately with `AbstractComponent` it's still not working, but yeah, maybe that `type Composed` is defined incorrectly. But on the other hand this type is written as an example in flow HOC docs, so perhaps this should be the recommended way...

And in tests for the PR there is no test which exports just the composition, just the HOC, without Component. Because in this scenario Flow wants to annotate exported `compose` function so there has to be done additional type definition.

Anyway, that's off-top i guess :)

**Comment by peter-leonov** [comment]
> And in tests for the PR there is no test which exports just the composition, just the HOC, without Component. Because in this scenario Flow wants to annotate exported `compose` function so there has to be done additional type definition.

Since Flow v.0.85 it is the intended behaviour. I'd say we have to deal with it 😎 First week I was trying to avoid as much type parameters as possible and Flow was 99% correct finding inconsistencies in my types but the error messages were not just incomplete they were horribly misleading making me spend 3 time more energy guessing where the real error was.

Currently I'm closely looking at Rust and they also do there bi-directional type inference aka OCaml (the initial implementation of Rust was written in OCaml as Flow still is) but the Rust team made a volitional decision and made it mandatory to specify type signatures for functions (not closures  though). Same in TypeScript. So, again, I guess, it's the new reality for all of us not using ML based languages.
